design/baremetal-operator/limit-hosts-provisioning.md

```diff
@@ -0,0 +1,120 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# limit-hosts-provisioning
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal aims to introduce a limit to the number of hosts that the
+BareMetal Operator can provision or register simultaneously as a safety
+measure to avoid overloading excessively the system
+
+## Motivation
+
+In the current implementation BMO takes care of handling the complete
+provisioning cycle for a bare metal host without considering the pressure
+level applied to the underlying Ironic services (DHCP included).
+A user that desires to scale up several hundreds of nodes in a single step
+could suffer issues produced from the sudden resources depletion
+
+### Goals
+
+1. Avoid any system instability due the excessive number of hosts being
+   provisioned simultaneously
+
+### Non-Goals
+
+1. Optmize the deployment time of large batches of nodes
+2. Allow the user to configure such behavior
+3. Modify the current Ironic service layout
+
+## Proposal
+
+BMO reconciliation loop already checks for Ironic availability before
+processing an update and the request gets rescheduled if Ironic is not ready.
+After that, a new check will verify if Ironic is overloaded or not so that
+the request could be rescheduled or processed as usual.
+
+### Implementation Details/Notes/Constraints
+
+The new check will verify the number of currently hosts being provisioned/
+registered and, based on a fixed threshold (initial propose is 20), will
+determine if the current update can be processed or needs to be rescheduled.
+The difficult part consists in determining the number of hosts currently being
+provisioned, since Ironic does not offer any way for querying about its
+workload. A couple of methods, to be discussed, are proposed in the subsequent
+sections to address this problem.
+
+#### Method 1
+
+Identify those nodes that already passed DHCP/PXE - considering it the most
+limiting factor for concurrent provisioning. The Ironic node’s field
+_driver\_internal\_info_ field can be analyzed for the presence of agent
+heartbeat (ie, for the presence of fields like _agent\_url_). A preliminary
+analysis revealed that this approach can have some drawbacks in case of pod
+restart: as soon as BMO comes back (along with the Ironic services) it will
+find already existing BMH resources as provisioned, and thus they will get
```

Comment:
Not sure it's a problem, we don't have to provide any throttling for or anyhow consider hosts that are already provisioned. By the way, normally provisioned hosts also may not have an `agent_url`. You need to limit your algorithm to Ironic nodes in certain states (`*wait*`, `*ing`).

Reply:
It sounds as a good approach. I'll update the proposal trying to enumerate the correct states.

Reply:
See previous comment